File: elixir/bob/bob.exs
defmodule Teenager do

  def hey(input) do
    cond do
      is_silent?(input) -> "Fine. Be that way!"
      is_question?(input) -> "Sure."
      is_yelling?(input) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end
  
  defp is_yelling?(input) do
    has_letters?(input) and input == String.upcase(input)
  end

  defp is_question? (input) do
    ends_in_question_mark?(input)
  end
  
  defp ends_in_question_mark?(input) do
    String.ends_with?(input, "?")
  end

  defp is_silent?(input) do
    not has_letters?(input) and not has_numbers?(input)
  end
  
  defp has_letters?(input) do
    String.match?(input, ~r/\p{L}/)
  end

  defp has_numbers?(input) do
    String.match?(input, ~r/\p{N}/)
  end
  
end

File: ocaml/bob/bob.ml
let is_question say =
  String.get say ((String.length say) - 1) == '?'

let is_shouting say =
  let up = String.uppercase say in
  (up = say) && (String.lowercase say <> up)

let is_empty say = say = ""
                           
let response_for say = let stripped_say = String.trim say in
                       match stripped_say with
                       | "" -> "Fine. Be that way!"
                       | said when (is_shouting said) -> "Whoa, chill out!"
                       | said when (is_question said) -> "Sure."
                       | _ -> "Whatever."

                                               
